rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,400 @@
+backward compatibility. The third part addresses new update strategy
+constraints between major Teleport versions.
+
+### Chart structure and deployed resources
+
+The resources in the chart would be split in two subdirectories,
+`templates/auth/` and `templates/proxy/` to clearly identify which resource is
+used by which teleport node. Common resources should be put in `templates/`.
+
+The chart would deploy two Deployments: one for the proxies and one for the auths.
+
+- the `teleport-proxy` Deployment: Those pods are stateless by default and can
+  be upscaled even in standalone mode. Deploying those nodes using a Deployment
+  means we cannot mount peristent storage on them. As Teleport does not support
+  graceful shutdown with record shipping, users might loose active sessions
+  recordings during a rollout if using the `proxy` mode. Teleport nodes
+  are relying on `kube` ProvisionTokens to join the auth nodes on startup ([see
+  RFD-0094](https://github.com/gravitational/teleport/blob/rfd/0096-helm-chart-revamp/rfd/0096-helm-chart-revamp.md)).
+- the `teleport-auth` Deployment:  Those pods cannot be
+  replicated without remote backend for state and audit logs. When persistence is
+  enabled, a single volume will be mounted to those pods and the update strategy
+  will be "re-create". For setups in which auth pods are stateless, the Deployment
+  can be scaled up.
+
+The main LB service should send traffic to the proxies, two additional services
+for in-cluster communication should be created: one for the proxies and one for
+the auth.
+
+The trust between auth and proxy should be bootstrapped by
+[creating a provisionToken on start](https://github.com/gravitational/teleport/pull/19009).
+
+#### Labels and selectors
+
+Deploying different pod sets requires a way to discriminate them. The only
+label set currently is `app: {{.Release.Name }}`. We should follow [Helm
+label recommendations](https://helm.sh/docs/chart_best_practices/labels/):
+
+| Label                        | Value                                 | Purpose   |
+|------------------------------|---------------------------------------|-----------|
+| app.kubernetes.io/name       | `{{- default .Chart.Name .Values.nameOverride \| trunc 63 \| trimSuffix "-" }}` | Identify the application. |
+| helm.sh/chart                | `{{ .Chart.Name }}-{{ .Chart.Version \| replace "+" "_" }}` | This should be the chart name and version. |
+| app.kubernetes.io/managed-by | `{{ .Release.Service }}`              | It is for finding all things managed by Helm. |
+| app.kubernetes.io/instance   | `{{ .Release.Name }}`                 | It aids in differentiating between different instances of the same application. |
+| app.kubernetes.io/version    | `{{ .Chart.AppVersion }}`             | The version of the app. |
+| app.kubernetes.io/component  | `auth` or `proxy`                     | This is a common label for marking the different roles that pieces may play in an application. |
```

Comment:
It seems like `app.kubernetes.io/component` should have a value that indicates which teleport components are active. auth+kube or proxy+kube, etc...

If not in that label, is there another label that would better indicate the teleport component(s) that are active?

It would make the helm chart feel more in tune with the product. It'd be great if this could be generated based on the rendered teleport.yaml in the configmap for the deployment, if possible.

Reply:
Fair, currently it would be "proxy" and "auth+kube". However I'd like to move kube service out of auth pods soon, I kept this off the main PR in a failed attempt to limit the number of changes. Eventually, this will be `proxy`, `auth`, and `kube`. I'll update the RFD to reflect this.

However I think this will be easier to just write `auth` instead of `auth+kube` today as we rely on those labels as selectors and Kubernetes does not support label edition well, this will require another downtime when we'll move the kube service out of the auth because of the label change.